File: 310971/Guzik/task-320/src/main/java/com/example/entitiesapp/services/WriterService.java
package com.example.entitiesapp.services;

import com.example.entitiesapp.dto.request.WriterRequestTo;
import com.example.entitiesapp.dto.response.WriterResponseTo;
import com.example.entitiesapp.entities.Writer;
import com.example.entitiesapp.exceptions.DuplicateResourceException;
import com.example.entitiesapp.exceptions.ResourceNotFoundException;
import com.example.entitiesapp.exceptions.ValidationException;
import com.example.entitiesapp.mappers.WriterMapper;
import com.example.entitiesapp.repositories.WriterRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class WriterService {
    private final WriterRepository writerRepository;
    private final WriterMapper writerMapper;

    public List<WriterResponseTo> getAll() {
        return writerRepository.findAll().stream()
                .map(writerMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    public WriterResponseTo getById(Long id) {
        Writer writer = writerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Writer not found with id: " + id));
        return writerMapper.toResponseDto(writer);
    }

    @Transactional
    public WriterResponseTo create(WriterRequestTo dto) {
        validateWriterRequest(dto);

        if (writerRepository.existsByLogin(dto.getLogin())) {
            throw new DuplicateResourceException("Writer with login '" + dto.getLogin() + "' already exists");
        }

        Writer writer = writerMapper.toEntity(dto);
        Writer saved = writerRepository.save(writer);
        return writerMapper.toResponseDto(saved);
    }

    @Transactional
    public WriterResponseTo update(Long id, WriterRequestTo dto) {
        validateWriterRequest(dto);

        Writer existing = writerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Writer not found with id: " + id));

        if (!existing.getLogin().equals(dto.getLogin()) &&
                writerRepository.existsByLogin(dto.getLogin())) {
            throw new DuplicateResourceException("Writer with login '" + dto.getLogin() + "' already exists");
        }

        existing.setLogin(dto.getLogin());
        existing.setPassword(dto.getPassword());
        existing.setFirstName(dto.getFirstname());
        existing.setLastName(dto.getLastname());

        Writer updated = writerRepository.save(existing);
        return writerMapper.toResponseDto(updated);
    }

    @Transactional
    public void delete(Long id) {
        if (!writerRepository.existsById(id)) {
            throw new ResourceNotFoundException("Writer not found with id: " + id);
        }

        Writer writer = writerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Writer not found with id: " + id));

        if (!writer.getArticles().isEmpty()) {
            throw new ValidationException("Cannot delete writer with id " + id +
                    " because they have " + writer.getArticles().size() + " articles");
        }

        writerRepository.deleteById(id);
    }

    private void validateWriterRequest(WriterRequestTo dto) {
        if (dto.getLogin() == null || dto.getLogin().trim().isEmpty()) {
            throw new ValidationException("Login is required");
        }
        if (dto.getLogin().length() < 2 || dto.getLogin().length() > 64) {
            throw new ValidationException("Login must be between 2 and 64 characters");
        }
        if (dto.getPassword() == null || dto.getPassword().trim().isEmpty()) {
            throw new ValidationException("Password is required");
        }
        if (dto.getPassword().length() < 8 || dto.getPassword().length() > 128) {
            throw new ValidationException("Password must be between 8 and 128 characters");
        }
        if (dto.getFirstname() == null || dto.getFirstname().trim().isEmpty()) {
            throw new ValidationException("Firstname is required");
        }
        if (dto.getFirstname().length() < 2 || dto.getFirstname().length() > 64) {
            throw new ValidationException("Firstname must be between 2 and 64 characters");
        }
        if (dto.getLastname() == null || dto.getLastname().trim().isEmpty()) {
            throw new ValidationException("Lastname is required");
        }
        if (dto.getLastname().length() < 2 || dto.getLastname().length() > 64) {
            throw new ValidationException("Lastname must be between 2 and 64 characters");
        }
    }

    public boolean existsById(Long id) {
        return writerRepository.existsById(id);
    }

    public WriterResponseTo findByLogin(String login) {
        Writer writer = writerRepository.findByLogin(login)
                .orElseThrow(() -> new ResourceNotFoundException("Writer not found with login: " + login));
        return writerMapper.toResponseDto(writer);
    }
}

File: 351002/Bubich/t310/app/services/writer_service.py
from typing import List
from app.models.writer import Writer
from app.dto.requests.writer_request import WriterRequestTo
from app.dto.responses.writer_response import WriterResponseTo
from app.repository.in_memory_repository import InMemoryRepository
from app.exceptions.custom_exceptions import NotFoundError, ValidationError


class WriterService:
    def __init__(self, repository: InMemoryRepository):
        self.repository = repository

    def create(self, request: WriterRequestTo) -> WriterResponseTo:
        self._validate_request(request)
        writer = Writer(
            login=request.login,
            password=request.password,
            firstname=request.firstname,
            lastname=request.lastname
        )
        saved_writer = self.repository.save(writer)
        return self._to_response(saved_writer)

    def get_by_id(self, id: int) -> WriterResponseTo:
        writer = self.repository.find_by_id(id)
        if not writer:
            raise NotFoundError(f"Writer with id {id} not found")
        return self._to_response(writer)

    def get_all(self) -> List[WriterResponseTo]:
        writers = self.repository.find_all()
        return [self._to_response(w) for w in writers]

    def update(self, id: int, request: WriterRequestTo) -> WriterResponseTo:
        self._validate_request(request)
        existing_writer = self.repository.find_by_id(id)
        if not existing_writer:
            raise NotFoundError(f"Writer with id {id} not found")

        existing_writer.login = request.login
        existing_writer.password = request.password
        existing_writer.firstname = request.firstname
        existing_writer.lastname = request.lastname

        updated_writer = self.repository.update(existing_writer)
        return self._to_response(updated_writer)

    def delete(self, id: int) -> None:
        if not self.repository.delete_by_id(id):
            raise NotFoundError(f"Writer with id {id} not found")

    def get_by_story_id(self, story_repo: InMemoryRepository, story_id: int) -> WriterResponseTo:
        story = story_repo.find_by_id(story_id)
        if not story:
            raise NotFoundError(f"Story with id {story_id} not found")

        writer = self.repository.find_by_id(story.writer_id)
        if not writer:
            raise NotFoundError(f"Writer for story {story_id} not found")

        return self._to_response(writer)

    def _validate_request(self, request: WriterRequestTo):
        if not request.login or len(request.login) < 2 or len(request.login) > 64:
            raise ValidationError("Login must be between 2 and 64 characters")
        if not request.password or len(request.password) < 8 or len(request.password) > 128:
            raise ValidationError("Password must be between 8 and 128 characters")
        if not request.firstname or len(request.firstname) < 2 or len(request.firstname) > 64:
            raise ValidationError("Firstname must be between 2 and 64 characters")
        if not request.lastname or len(request.lastname) < 2 or len(request.lastname) > 64:
            raise ValidationError("Lastname must be between 2 and 64 characters")

    def _to_response(self, writer: Writer) -> WriterResponseTo:
        return WriterResponseTo(
            id=writer.id,
            login=writer.login,
            firstname=writer.firstname,
            lastname=writer.lastname
        )